### ISSUE TITLE
UILocalNotification

### ISSUE THREAD

**Issue by hjanuschka** [OneSignal/OneSignal-iOS-SDK#139]
I am trying to get UILocalNotification working. 
i tried all delegates i found. none of them is called when launching the app, or when the app is in background, app is foregrounded but no callback called.

is there an official guide on how to use onesignal with uilocalnotifications.
my remote ones are working fine, 

OneSignal Version: 2.2.3

**Comment by hjanuschka** [comment]
ok i kinda found - atleast a workaround:

i tried UILocalNotification, then i tried UNNotitfications, nothing worked, so at the end of the day it was:




```objective-c
UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
center.delegate=self;


[OneSignal initWithLaunchOptions:launchOptions appId:@"XXXXXXXXX"     handleNotificationReceived:^(OSNotification *notification) {

        } handleNotificationAction:^(OSNotificationOpenedResult *result) {
            OSNotificationPayload* payload = result.notification.payload;
            if (payload.additionalData) {
                [self managePush:payload.additionalData];
            }
            
} settings:nil];
```

works and final my:

```objective-c
- (void) userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)())completionHandler {
```
is called  - but the `handleNotificationAction` is not called anymore in the OneSignal init block.
the underlying error seems to be - clicking/touching on a notification (local, unn).


```
2016-11-23 21:40:30.244 Krone[23014:63992680] Warning: UNUserNotificationCenter delegate received call to -userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler: but the completion handler was never called.
```

i also tried looking at the source, i suspect it somwhere here: https://github.com/OneSignal/OneSignal-iOS-SDK/blob/d0241ba6010b3eb6d35a3cdadc465945a08dd422/iOS_SDK/OneSignal/UNUserNotificationCenter%2BOneSignal.m#L125


can someone help me with that, as the above "workaround" locks the whole stuff to ios10+

**Comment by hjanuschka** [comment]
the best solution would be if i can stick to this:

```objective-c
                    UILocalNotification *notification = [[UILocalNotification alloc] init];
                    notification.alertBody = responseObject[@"title"];
                    notification.soundName = UILocalNotificationDefaultSoundName;
                    
                    [[UIApplication sharedApplication] presentLocalNotificationNow:notification];

```


and https://github.com/OneSignal/OneSignal-iOS-SDK/blob/d0241ba6010b3eb6d35a3cdadc465945a08dd422/iOS_SDK/OneSignal/UIApplicationDelegate%2BOneSignal.m#L117 should handle all the cases as it seems it's supposed to be 9,10 support.

**Comment by jkasten2** [comment]
@hjanuschka I recommend the same. It the the least complex setup to use the old iOS 9 methods and delete selectors such as `application:didReceiveLocalNotification:` as they also fire for iOS 10.

Once you set a `UNUserNotificationCenter` delegate on iOS 10 it assumes you have updated all your code to use the new selector and will no longer fire any of the 6 notification open/received events.

The behavior noted above happens even without our SDK.

Our `handleNotificationReceived` and `handleNotificationAction` blocks do not fire for non-OneSignal notifications or local notifications created by your app's code.

Let us know if this covers everything and I can close this issue.

Thanks.

**Comment by hjanuschka** [comment]
@jkasten2 thx for reply - i tried with - UILocalNotification but no delegates are fire.
do you have an idea?

**Comment by hjanuschka** [comment]
@jkasten2  ok i made some progress.


```objective-c

- (void) application:(UIApplication *)application didReceiveLocalNotification:(UILocalNotification *)notification {
    [self managePush:notification.userInfo];
}
- (void) userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)())completionHandler {
    [self managePush:response.notification.request.content.userInfo];
    if(completionHandler) {
        completionHandler();
    }
}



- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {

   [OneSignal initWithLaunchOptions:launchOptions appId:@"XXXXXXXXX" handleNotificationReceived:^(OSNotification *notification) {
        } handleNotificationAction:^(OSNotificationOpenedResult *result) {
            OSNotificationPayload* payload = result.notification.payload;
            
            
            if (payload.additionalData) {
                [self managePush:payload.additionalData];
            }
            
        } settings:nil];
        if(SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"10.0")) {
            UNUserNotificationCenter * center = [UNUserNotificationCenter currentNotificationCenter];
            center.delegate=self;
        }
        if(SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"8.0")) {
            if ([application respondsToSelector:@selector(registerUserNotificationSettings:)]) {
                [application registerUserNotificationSettings:[UIUserNotificationSettings settingsForTypes:UIUserNotificationTypeAlert|UIUserNotificationTypeBadge|UIUserNotificationTypeSound categories:nil]];
            }
        }

}
```



this works with ios8/9/10.


on 10 - i override the delegate - which then calls `userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)())completionHandler` -> disabling the OneSignal Block - but i think its fine.


on 8 and 9:

the `application:(UIApplication *)application didReceiveLocalNotification:(UILocalNotification *)notification` is getting called.


can you confirm that this is ok from your side, from a point of device registration view.

**Comment by jkasten2** [comment]
@hjanuschka Thanks for your details. I have reproduced your issue and have fixed it in commit 543b490cd9bb94143d5558106b4fce26588f3c8e.

```Objective-C
UILocalNotification *notification = [[UILocalNotification alloc] init];
notification.alertBody =@"title";
notification.soundName = UILocalNotificationDefaultSoundName;
                    
[[UIApplication sharedApplication] presentLocalNotificationNow:notification];
```

This will then trigger `application:didReceiveLocalNotification:` as it should in the next SDK release.

Feel free to reopen this issue if you continue to see a problem after this.

**Comment by hjanuschka** [comment]
@jkasten2  awesome, any ETA when the sdk will be released?
